vltt/assignment3/solution.cpp

```diff
@@ -18,180 +17,226 @@
   } else {                                                                \
     std::cout << __FILE__ << ":" << __LINE__ << " True " << std::endl;  \
   }
-  
 
 template <class T>
 inline std::ostream& operator<<(std::ostream& os, const std::set<T>& s) {
-    os << "{";
-    for (auto &it : s) {
-        os << it;
-        std::cout << ", " ;
-    }
-    os << "}";
-    return os;
+	os << "{";
+	for (auto &it : s) {
+		os << it;
+		std::cout << ", ";
+	}
+	os << "}";
+	return os;
 }
 
 class Dictionary {
 private:
-    class Trie {
-    private:
-        class Node {
-        private:
-            std::map<char, Node*> go;
-        public:
-            bool terminal = false;
-            void add(const char &letter) {
-                go[letter] = new Node;
-            }
-            bool isHere(const char &letter) {
-                return go.find(letter) != go.end();
-            }
-            Node *next(const char &letter) {
-                if (go.find(letter) == go.end()) return nullptr;
-                return go[letter];
-            }
-        };
-        Node *start;
-    public:
-        Trie() {
-            start = new Node;
-        }
-        //add new word to the Trie
-        void add(const std::string &word) {
-            if (word.empty()) return;
-            Node *cur = start;
-            for (int i = 0; i < word.length(); i++) {
-                char letter = word[i];
-                if (cur->isHere(letter) == false) {
-                    cur->add(letter);
-                }
-                cur = cur->next(letter);
-                if (i == int(word.length()) - 1) {
-                    cur->terminal = true;
-                }
-            }
-        }
-        //determines whether there is such a prefix in the Trie
-        //return: -1 - prefix, 0 - nothing, 1 - word
-        int find_str(const std::string &str) {
-            if (str.empty()) return 0;
-            Node *cur = start;
-            for (int i = 0; i < str.length(); i++) {
-                if (cur == nullptr) break;
-                char letter = str[i];
-                cur = cur->next(letter);
-                if (i == int(str.length()) - 1) {
-                    if (cur == nullptr) return 0;
-                    if (cur->terminal == true) return 1;
-                    else return -1;
-                }
-            }
-            return 0;
-        }
-    };
-    Trie prefixes;
+	class Trie {
+	private:
+		class Node {
+		private:
+			std::map<char, std::shared_ptr<Node>> children;
+		public:
+			bool terminal = false;
```

Comment:
I can not make it private because I use it in class Trie.

Reply:
Ah, I see! It's fine the way it is then :)